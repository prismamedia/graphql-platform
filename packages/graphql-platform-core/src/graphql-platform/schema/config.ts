import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
