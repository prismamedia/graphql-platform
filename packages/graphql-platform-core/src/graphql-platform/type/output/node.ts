import {
  GraphQLFieldConfig,
  GraphQLNonNullDecorator,
  GraphQLSelectionNode,
  MaybePromise,
  parseGraphQLResolveInfo,
  POJO,
  SuperMapOfNamedObject,
} from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import {
  GraphQLFieldConfigMap,
  GraphQLObjectType,
  GraphQLResolveInfo,
} from 'graphql';
import { AnyBaseContext, BaseContext } from '../../../graphql-platform';
import {
  Field,
  FieldValue,
  InverseRelation,
  Relation,
  VirtualField,
} from '../../resource';
import { AbstractOutputType } from '../abstract-type';

export enum NodeFieldKind {
  Field,
  Relation,
  InverseRelation,
  InverseRelationCount,
  VirtualField,
}

export type NodeField = {
  name: string;
  public: boolean;
  config: GraphQLFieldConfig;
} & (
  | {
      kind: NodeFieldKind.Field;
      field: Field;
    }
  | {
      kind: NodeFieldKind.Relation;
      relation: Relation;
    }
  | {
      kind: NodeFieldKind.InverseRelation;
      inverseRelation: InverseRelation;
    }
  | {
      kind: NodeFieldKind.InverseRelationCount;
      inverseRelation: InverseRelation;
    }
  | {
      kind: NodeFieldKind.VirtualField;
      virtualField: VirtualField;
    }
);

export class NodeFieldMap extends SuperMapOfNamedObject<NodeField> {}

export type NodeSourceFieldValue =
  | FieldValue
  | NodeSource
  | NodeSource[]
  | number;

export type NodeSourceFieldResolverParams<
  TArgs extends POJO = any,
  TContext extends AnyBaseContext = any
> = Readonly<{
  args: TArgs;
  context: TContext;
  selectionNode: GraphQLSelectionNode;
}>;

export type NodeSourceFieldResolver<
  TArgs extends POJO = any,
  TContext extends AnyBaseContext = any
> = (
  params: NodeSourceFieldResolverParams<TArgs, TContext>,
) => MaybePromise<NodeSourceFieldValue>;

export type NodeSourceField = NodeSourceFieldValue | NodeSourceFieldResolver;

export type NodeSource = {
  [fieldName: string]: NodeSourceField;
};

export class NodeType extends AbstractOutputType {
  @Memoize()
  public get name() {
    return this.resource.name;
  }

  @Memoize()
  public get description() {
    return this.resource.description || `"${this.resource}" resource's node`;
  }

  protected async getSourceFieldValue(
    source: NodeSource,
    args: POJO,
    context: BaseContext,
    info: GraphQLResolveInfo,
  ): Promise<NodeSourceFieldValue> {
    const selectionNode = parseGraphQLResolveInfo(info);
    const fieldValue = source[selectionNode.name];

    return typeof fieldValue === 'function'
      ? fieldValue({ args, context, selectionNode })
      : fieldValue;
  }

  @Memoize()
  public getFieldMap(): NodeFieldMap {
    const nodeFieldMap = new NodeFieldMap();

    for (const field of this.resource.getFieldSet()) {
      const nodeField: NodeField = {
        name: field.name,
        public: field.isPublic(),
        kind: NodeFieldKind.Field,
        field,
        config: {
          description: field.description,
          type: GraphQLNonNullDecorator(field.getType(), !field.isNullable()),
          resolve: this.getSourceFieldValue,
        },
      };

      nodeFieldMap.set(nodeField.name, nodeField);
    }

    for (const relation of this.resource.getRelationSet()) {
      const findOne = relation.getTo().getQuery('FindOne');

      const nodeField: NodeField = {
        name: relation.name,
        public: relation.isPublic() && findOne.isPublic(),
        kind: NodeFieldKind.Relation,
        relation,
        config: {
          description: relation.description,
          type: GraphQLNonNullDecorator(
            findOne.getGraphQLFieldConfigType(),
            !relation.isNullable(),
          ),
          resolve: this.getSourceFieldValue,
        },
      };

      nodeFieldMap.set(nodeField.name, nodeField);
    }

    for (const inverseRelation of this.resource.getInverseRelationSet()) {
      if (inverseRelation.isToOne()) {
        // Find one
        const findOne = inverseRelation.getTo().getQuery('FindOne');

        const nodeField: NodeField = {
          name: inverseRelation.name,
          public: inverseRelation.relation.isPublic() && findOne.isPublic(),
          kind: NodeFieldKind.InverseRelation,
          inverseRelation,
          config: {
            description: inverseRelation.description,
            type: findOne.getGraphQLFieldConfigType(),
            resolve: this.getSourceFieldValue,
          },
        };

        nodeFieldMap.set(nodeField.name, nodeField);
      } else {
        {
          // Find many
          const findMany = inverseRelation.getTo().getQuery('FindMany');

          const nodeField: NodeField = {
            name: inverseRelation.name,
            public: inverseRelation.relation.isPublic() && findMany.isPublic(),
            kind: NodeFieldKind.InverseRelation,
            inverseRelation,
            config: {
              description: inverseRelation.description,
              args: findMany.getGraphQLFieldConfigArgs(),
              type: findMany.getGraphQLFieldConfigType(),
              resolve: this.getSourceFieldValue,
            },
          };

          nodeFieldMap.set(nodeField.name, nodeField);
        }

        {
          // Count
          const count = inverseRelation.getTo().getQuery('Count');

          const nodeField: NodeField = {
            name: inverseRelation.countName,
            public: inverseRelation.relation.isPublic() && count.isPublic(),
            kind: NodeFieldKind.InverseRelationCount,
            inverseRelation,
            config: {
              description: inverseRelation.countDescription,
              args: count.getGraphQLFieldConfigArgs(),
              type: count.getGraphQLFieldConfigType(),
              resolve: this.getSourceFieldValue,
            },
          };

          nodeFieldMap.set(nodeField.name, nodeField);
        }
      }
    }

    for (const virtualField of this.resource.getVirtualFieldSet()) {
      const nodeField: NodeField = {
        name: virtualField.name,
        public: true,
        kind: NodeFieldKind.VirtualField,
        virtualField,
        config: virtualField.config,
      };

      nodeFieldMap.set(nodeField.name, nodeField);
    }

    return nodeFieldMap;
  }

  @Memoize()
  public getGraphQLType() {
    return new GraphQLObjectType({
      name: this.name,
      description: this.description,
      fields: () => {
        const fields: GraphQLFieldConfigMap<NodeSource, BaseContext> = {};

        for (const field of this.getFieldMap().values()) {
          if (field.public) {
            fields[field.name] = field.config;
          }
        }

        return fields;
      },
    });
  }
}
