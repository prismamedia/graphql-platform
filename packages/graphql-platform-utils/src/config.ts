export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
