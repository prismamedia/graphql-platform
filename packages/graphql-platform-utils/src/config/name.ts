import * as graphql from 'graphql';
import { UnexpectedValueError } from '../error.js';
import type { Path } from '../path.js';

/**
 * A string compliant with the GraphQL specification
 *
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export type Name = string;

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function assertName(
  maybeName: unknown,
  path: Path,
): asserts maybeName is Name {
  if (typeof maybeName !== 'string' || !maybeName) {
    throw new UnexpectedValueError(`a non-empty string`, maybeName, {
      path,
    });
  }

  try {
    graphql.assertName(maybeName);
  } catch (error) {
    throw new UnexpectedValueError(
      'to be valid against the GraphQL "Names" specification (@see: https://spec.graphql.org/draft/#sec-Names)',
      maybeName,
      { cause: error, path },
    );
  }
}

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function ensureName(maybeName: unknown, path: Path): Name {
  assertName(maybeName, path);

  return maybeName;
}
