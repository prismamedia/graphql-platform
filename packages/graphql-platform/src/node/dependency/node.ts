import * as utils from '@prismamedia/graphql-platform-utils';
import assert from 'node:assert';
import type { Node } from '../../node.js';
import { NodeCreation, NodeDeletion, type NodeChange } from '../change.js';
import type { Component } from '../definition.js';
import { MutationContextChangesByNode } from '../operation/mutation/context/changes.js';

export type NodeDependencyJSON = {
  [utils.MutationType.CREATION]?: boolean;
  [utils.MutationType.DELETION]?: boolean;
  [utils.MutationType.UPDATE]?: Component['name'][];
};

export class NodeDependency {
  public readonly [utils.MutationType.CREATION]: boolean;
  public readonly [utils.MutationType.DELETION]: boolean;
  public readonly [utils.MutationType.UPDATE]: ReadonlySet<Component>;

  public readonly normalized?: this;

  public constructor(
    public readonly node: Node,
    config?: {
      [utils.MutationType.CREATION]?: boolean;
      [utils.MutationType.DELETION]?: boolean;
      [utils.MutationType.UPDATE]?: ReadonlySet<Component>;
    },
  ) {
    this[utils.MutationType.CREATION] =
      config?.[utils.MutationType.CREATION] === true;

    this[utils.MutationType.DELETION] =
      config?.[utils.MutationType.DELETION] === true;

    this[utils.MutationType.UPDATE] = new Set(
      config?.[utils.MutationType.UPDATE],
    );

    this.normalized =
      this[utils.MutationType.CREATION] ||
      this[utils.MutationType.DELETION] ||
      this[utils.MutationType.UPDATE].size
        ? this
        : undefined;
  }

  public mergeWith(other?: NodeDependency): this | NodeDependency {
    if (!other) {
      return this;
    }

    assert.strictEqual(other.node, this.node);

    return new NodeDependency(this.node, {
      [utils.MutationType.CREATION]:
        this[utils.MutationType.CREATION] || other[utils.MutationType.CREATION],

      [utils.MutationType.DELETION]:
        this[utils.MutationType.DELETION] || other[utils.MutationType.DELETION],

      [utils.MutationType.UPDATE]: this[utils.MutationType.UPDATE].union(
        other[utils.MutationType.UPDATE],
      ),
    });
  }

  public equals(other: NodeDependency): boolean {
    return utils.mutationTypes.every((mutationType) => {
      switch (mutationType) {
        case utils.MutationType.UPDATE:
          return (
            this[mutationType].size === other[mutationType].size &&
            !this[mutationType].difference(other[mutationType]).size
          );

        default:
          return this[mutationType] === other[mutationType];
      }
    });
  }

  public dependsOn(change: NodeChange): boolean {
    return change instanceof NodeCreation
      ? this[utils.MutationType.CREATION]
      : change instanceof NodeDeletion
        ? this[utils.MutationType.DELETION]
        : this[utils.MutationType.UPDATE].size && change.updatesByComponent.size
          ? !this[utils.MutationType.UPDATE].isDisjointFrom(
              change.updatesByComponent,
            )
          : false;
  }

  public dependsOnChanges(changes?: MutationContextChangesByNode): boolean {
    return changes?.size
      ? utils.mutationTypes.some((mutationType) => {
          switch (mutationType) {
            case utils.MutationType.UPDATE:
              return this[mutationType].size
                ? changes[mutationType]
                    .values()
                    .some(
                      ({ updatesByComponent }) =>
                        !this[mutationType].isDisjointFrom(updatesByComponent),
                    )
                : false;

            default:
              return this[mutationType] && changes[mutationType].size > 0;
          }
        })
      : false;
  }

  public filterChanges<TRequestContext extends object>(
    changes?: MutationContextChangesByNode<TRequestContext>,
  ): IteratorObject<NodeChange<TRequestContext>> | undefined {
    if (!changes?.size) {
      return undefined;
    }

    return utils.mutationTypes.values().flatMap<NodeChange>((mutationType) => {
      switch (mutationType) {
        case utils.MutationType.UPDATE:
          return this[mutationType].size
            ? changes[mutationType]
                .values()
                .filter(
                  ({ updatesByComponent }) =>
                    !this[mutationType].isDisjointFrom(updatesByComponent),
                )
            : [];

        default:
          return this[mutationType] ? changes[mutationType].values() : [];
      }
    });
  }

  public toJSON(): NodeDependencyJSON {
    return {
      ...(this[utils.MutationType.CREATION] && {
        [utils.MutationType.CREATION]: this[utils.MutationType.CREATION],
      }),

      ...(this[utils.MutationType.DELETION] && {
        [utils.MutationType.DELETION]: this[utils.MutationType.DELETION],
      }),

      ...(this[utils.MutationType.UPDATE].size && {
        [utils.MutationType.UPDATE]: Array.from(
          this[utils.MutationType.UPDATE],
          ({ name }) => name,
        ),
      }),
    };
  }
}
