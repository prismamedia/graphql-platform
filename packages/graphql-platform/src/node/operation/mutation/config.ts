import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
