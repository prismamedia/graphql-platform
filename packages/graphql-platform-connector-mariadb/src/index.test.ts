import { MyGP } from '@prismamedia/graphql-platform/__tests__/config.js';
import { EOL } from 'node:os';
import {
  AddTableForeignKeysStatement,
  CreateSchemaStatement,
  CreateTableStatement,
  DropSchemaStatement,
  MariaDBConnector,
} from './index.js';
import { createGraphQLPlatform } from './__tests__/config.js';

describe('GraphQL-Platform Connector MariaDB', () => {
  let gp: MyGP<MariaDBConnector>;

  beforeAll(async () => {
    gp = createGraphQLPlatform('connector_mariadb');
  });

  it('generates valid and stable schema', async () => {
    expect(
      new DropSchemaStatement(gp.connector.schema, { ifExists: true }).sql,
    ).toMatchSnapshot();

    expect(
      new CreateSchemaStatement(gp.connector.schema, { orReplace: true }).sql,
    ).toMatchSnapshot();

    expect(
      Array.from(
        gp.connector.schema.tablesByNode.values(),
        (table) => new CreateTableStatement(table).sql,
      ).join(EOL.repeat(2)),
    ).toMatchSnapshot();

    expect(
      Array.from(gp.connector.schema.tablesByNode.values(), (table) =>
        table.foreignKeysByEdge.size
          ? new AddTableForeignKeysStatement(table).sql
          : undefined,
      )
        .filter(Boolean)
        .join(EOL.repeat(2)),
    ).toMatchSnapshot();
  });
});
