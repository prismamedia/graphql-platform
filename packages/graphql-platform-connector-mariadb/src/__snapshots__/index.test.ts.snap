// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL-Platform Connector MariaDB generates valid and stable schema 1`] = `"DROP SCHEMA IF EXISTS \`tests_connector_mariadb\`"`;

exports[`GraphQL-Platform Connector MariaDB generates valid and stable schema 2`] = `
"CREATE OR REPLACE SCHEMA \`tests_connector_mariadb\`
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'"
`;

exports[`GraphQL-Platform Connector MariaDB generates valid and stable schema 3`] = `
"CREATE TABLE \`tests_connector_mariadb\`.\`articles\` (
  \`private_id\` INT UNSIGNED AUTO_INCREMENT NOT NULL COMMENT 'This id is used to identify an Article internally',
  \`id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL COMMENT 'This UUID identifies an Article publicly',
  \`status\` ENUM('draft','published','deleted') NOT NULL,
  \`title\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`slug\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`body\` JSON NULL COMMENT 'The article\\'s body',
  \`category_private_id\` INT UNSIGNED NULL,
  \`created_by_id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  \`created_at\` DATETIME(3) NOT NULL,
  \`updated_by_username\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  \`updated_at\` TIMESTAMP(0) NOT NULL,
  \`metas\` JSON NULL COMMENT 'Contains any arbitrary data you want to store alongside the article',
  \`highlighted\` BOOLEAN NULL COMMENT 'Is the article highlighted?',
  \`sponsored\` BOOLEAN NULL COMMENT 'Is the article a partnership?',
  \`views\` BIGINT UNSIGNED NOT NULL,
  \`score\` DOUBLE(10,2) UNSIGNED NOT NULL,
  \`machine_tags\` JSON NULL,
  PRIMARY KEY (\`private_id\`),
  UNIQUE INDEX \`unq_id\` (\`id\`),
  UNIQUE INDEX \`unq_category_private_id_slug\` (\`category_private_id\`,\`slug\`),
  FULLTEXT \`ft_title\` (\`title\`),
  INDEX \`idx_slug\` (\`slug\`),
  INDEX \`idx_status_slug\` (\`status\`,\`slug\`),
  INDEX \`idx_category_private_id_updated_at\` (\`category_private_id\`,\`updated_at\`),
  FOREIGN KEY \`fk_articles_category_private_id\` (\`category_private_id\`) REFERENCES \`categories\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  FOREIGN KEY \`fk_articles_created_by_id\` (\`created_by_id\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  FOREIGN KEY \`fk_articles_updated_by_username\` (\`updated_by_username\`) REFERENCES \`users\` (\`username\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'
COMMENT 'The article is the main resource, written by the journalists'

CREATE TABLE \`tests_connector_mariadb\`.\`article_extensions\` (
  \`article_private_id\` INT UNSIGNED NOT NULL,
  \`source\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL COMMENT 'The source of the article',
  PRIMARY KEY (\`article_private_id\`),
  FOREIGN KEY \`fk_article_extensions_article_private_id\` (\`article_private_id\`) REFERENCES \`articles\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`categories\` (
  \`private_id\` INT UNSIGNED AUTO_INCREMENT NOT NULL,
  \`id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`title\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`slug\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`parent_private_id\` INT UNSIGNED NULL,
  \`order\` INT UNSIGNED NOT NULL,
  PRIMARY KEY (\`private_id\`),
  UNIQUE INDEX \`unq_id\` (\`id\`),
  UNIQUE INDEX \`unq_parent_private_id_slug\` (\`parent_private_id\`,\`slug\`),
  UNIQUE INDEX \`unq_parent_private_id_order\` (\`parent_private_id\`,\`order\`),
  FOREIGN KEY \`fk_categories_parent_private_id\` (\`parent_private_id\`) REFERENCES \`categories\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`tags\` (
  \`id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`deprecated\` BOOLEAN NULL COMMENT 'A tag can be deprecated',
  \`title\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`slug\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`created_at\` DATETIME(3) NOT NULL,
  \`updated_at\` DATETIME(3) NOT NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE INDEX \`unq_slug\` (\`slug\`)
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`article_tags\` (
  \`article_private_id\` INT UNSIGNED NOT NULL,
  \`tag_id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`order\` INT UNSIGNED NOT NULL,
  PRIMARY KEY (\`article_private_id\`,\`tag_id\`),
  UNIQUE INDEX \`unq_article_private_id_order\` (\`article_private_id\`,\`order\`),
  FOREIGN KEY \`fk_article_tags_article_private_id\` (\`article_private_id\`) REFERENCES \`articles\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  FOREIGN KEY \`fk_article_tags_tag_id\` (\`tag_id\`) REFERENCES \`tags\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`article_tag_moderations\` (
  \`theArticlePrivateId\` INT UNSIGNED NOT NULL,
  \`theTagId\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`moderator_id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`moderation\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  PRIMARY KEY (\`theArticlePrivateId\`,\`theTagId\`,\`moderator_id\`),
  FOREIGN KEY \`my_custom_fk_name\` (\`theArticlePrivateId\`,\`theTagId\`) REFERENCES \`article_tags\` (\`article_private_id\`,\`tag_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  FOREIGN KEY \`fk_article_tag_moderations_moderator_id\` (\`moderator_id\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`users\` (
  \`id\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`username\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`created_at\` DATETIME(3) NOT NULL,
  \`last_logged_in_at\` TIMESTAMP(0) NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE INDEX \`unq_username\` (\`username\`)
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`user_profiles\` (
  \`theUserId\` CHAR(36) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`birthday\` DATE NULL,
  \`facebook_id\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  \`google_id\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  \`twitter_handle\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NULL,
  PRIMARY KEY (\`theUserId\`),
  FOREIGN KEY \`fk_user_profiles_theUserId\` (\`theUserId\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'

CREATE TABLE \`tests_connector_mariadb\`.\`logs\` (
  \`private_id\` INT UNSIGNED AUTO_INCREMENT NOT NULL,
  \`message\` VARCHAR(255) CHARSET 'utf8mb4' COLLATE 'utf8mb4_unicode_520_ci' NOT NULL,
  \`created_at\` DATETIME(3) NOT NULL,
  PRIMARY KEY (\`private_id\`)
)
ENGINE = 'InnoDB'
DEFAULT CHARSET = 'utf8mb4'
DEFAULT COLLATE = 'utf8mb4_unicode_520_ci'"
`;

exports[`GraphQL-Platform Connector MariaDB generates valid and stable schema 4`] = `
"ALTER TABLE \`tests_connector_mariadb\`.\`articles\`
  ADD FOREIGN KEY \`fk_articles_category_private_id\` (\`category_private_id\`) REFERENCES \`categories\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  ADD FOREIGN KEY \`fk_articles_created_by_id\` (\`created_by_id\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  ADD FOREIGN KEY \`fk_articles_updated_by_username\` (\`updated_by_username\`) REFERENCES \`users\` (\`username\`) ON UPDATE RESTRICT ON DELETE RESTRICT

ALTER TABLE \`tests_connector_mariadb\`.\`article_extensions\`
  ADD FOREIGN KEY \`fk_article_extensions_article_private_id\` (\`article_private_id\`) REFERENCES \`articles\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT

ALTER TABLE \`tests_connector_mariadb\`.\`categories\`
  ADD FOREIGN KEY \`fk_categories_parent_private_id\` (\`parent_private_id\`) REFERENCES \`categories\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT

ALTER TABLE \`tests_connector_mariadb\`.\`article_tags\`
  ADD FOREIGN KEY \`fk_article_tags_article_private_id\` (\`article_private_id\`) REFERENCES \`articles\` (\`private_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  ADD FOREIGN KEY \`fk_article_tags_tag_id\` (\`tag_id\`) REFERENCES \`tags\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT

ALTER TABLE \`tests_connector_mariadb\`.\`article_tag_moderations\`
  ADD FOREIGN KEY \`my_custom_fk_name\` (\`theArticlePrivateId\`,\`theTagId\`) REFERENCES \`article_tags\` (\`article_private_id\`,\`tag_id\`) ON UPDATE RESTRICT ON DELETE RESTRICT,
  ADD FOREIGN KEY \`fk_article_tag_moderations_moderator_id\` (\`moderator_id\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT

ALTER TABLE \`tests_connector_mariadb\`.\`user_profiles\`
  ADD FOREIGN KEY \`fk_user_profiles_theUserId\` (\`theUserId\`) REFERENCES \`users\` (\`id\`) ON UPDATE RESTRICT ON DELETE RESTRICT"
`;
